### DISCUSSION TITLE
C# return keyword appears to malfunction

### DISCUSSION THREAD

**Discussion by astrohart** [dotnet/csharplang#4589] [Q&A] [answered]
Have posted this on [the Code Project](https://www.codeproject.com/Questions/5298403/Csharp-return-keyword-not-exiting-the-method-but-r) and on Visual Studio Feedback.

Visual Studio version: Visual Studio 2019 Enterprise, v16.9.2
Windows version: Windows 10 Professional 64-bit Version 1909 (OS Build: 18363.1198)

I have no control over the Windows Updates as this code is on a VM.

Hi all,

I just recently posted this on the Visual Studio Feedback forums.  I am getting weird behavior from the C# `return` keyword.  My language version is C# 7.3, with .NET Framework 4.8.

As a programmer with deep experience in C-style languages, my understanding is that the `return` keyword should immediately stop executing the currently-running method: no passing GO, no collecting $200.

However, in my debugging session (and running executable), the following is occurring:

[Imgur Animated Gif ^](https://imgur.com/a/hFU8guM)

What is going on is, the debugger correctly activates a call to the `return` keyword and then jumps, not to the close curly brace of the containing method, but rather to the method's final statement, which is executed.  The next instruction executed is the close curly brace of the containing method.  I am encountering this behavior in other places of my code as well.  I do not think that the final statement of the method should be executed when the `return` keyword is invoked in the method body prior to the final statement.

Is there something not configured correctly, or am I simply stroking out? :-D

Thank you in advance for any suggestions and/or advice.

Sincerely Yours,

Brian Hart

**Comment by astrohart** [comment, marked as answer]
I am using [PostSharp](https://www.postsharp.net) v6.9.3 in my project. 

When I decorate the offending method with their `[Log(AttributeExclude = true)]` attribute, which suppresses weaving for this method, then the issue goes away. 

However, I would hate to have to think that it will be necessary to disable this for all my code's methods that early-return! I have filed a [bug report on their issue-tracking system](https://support.postsharp.net/request/28420-c-return-keyword-breaks-when-postsharp), and hopefully, the issue will be resolved soon.

**Reply by HaloFour** [reply]
The use of PostSharp here certainly explains the behavior.  PostSharp is rewriting your method to put code before/after your code, so when you "return" from your method you're actually jumping to the code that PostSharp weaved into your method to execute after your method is finished executing.  If you're using a logging aspect with PostSharp the behavior that you're describing is exactly the behavior that I would expect to see.

**Reply by HaloFour** [reply]
Let me take a step back, though.  The behavior I think you're seeing in the debugger is due to a mismatch in the debugging symbols emitted by the compiler and the binary rewritten by PostSharp.  It might look like the last line of the code you've written is being executed but I doubt that is the case.  That's something you might want to take up with PostSharp.

However, the behavior of something executing after `return` in a method is exactly the behavior I'd expect with an AOP weaver like PostSharp.  There are circumstances within the C# language where code within the method will execute after a `return`, like if that `return` is within a `try` block that has a `finally` clause.  However, in this case, you're using a tool that is rewriting the binary emitted by the C# compiler that is designed to wrap your code in additional logic.  I would assume that a logging aspect effectively wraps the body of your method in the IL equivalent of a `try`/`finally` so that it can inject logic into when your method exits, and as such it is expected that additional logic will execute within your method after your code returns.

**Reply by astrohart** [reply]
@HaloFour i think that I’m very much aware of how compile time code weaving works.   If it had not been for your explanation I might’ve gotten confused, and fallen down.  Sorry, I don’t mean to be brusque, but I think it’s safe to assume that people on this board know what they’re doing in terms of programming, as well as how the software works. Otherwise I have to quit my day job.

I think, rather, it’s a bug in the postsharp weaving implementation itself.  Their weaver should be smart enough to see  the IL and invoke the correct aspect functionality in the weave.  This is why I, in my own answer, I stated that I had filed a bug report with Post Sharp.

Funnily enough, I see your point about the debugger, but what I experienced was unusual behavior in the app itself while running. The method that checks whether the clear all operation can proceed shows the user a confirmation message box. If the user clicks the no button in the message box, then the check method returns false, And this would indicate that the operation may not continue. However even just running the app standalone and not in the debugger, My experience has been that it will simply ignore the click of the no button, and perform the clear operation regardless. This, of course, is inconsistent with and I was saying that the debugger and the code are out of sync.

**Reply by CyrusNajmabadi** [reply]
> Sorry, I don’t mean to be brusque, but I think it’s safe to assume that people on this board know what they’re doing in terms of programming, 

It's not actually safe to assume that.  People here come from all walks of life all all different technical levels.  Halo was simply trying to provide enough information to help out regardless of where you might be in that spectrum. :)

At this point though, i think this issue has run its course.  Even if this isn't an issue with your weaver (and is instead potentially an issue with VS debugger or the C# compiler), dotnet/csharplang is not the right venue for that conversation.  This is a repo centered around the design of hte language, not the surrounding tooling in the .net ecosystem.  Thanks! :)

**Reply by astrohart** [reply]
I think that it is safe to close this thread. Since this board is a C-sharp forum, And I think it’s safe to say the issue identify it does not seem endemic to the C-sharp compiler, and then this thread is no longer relevant.

**Reply by CyrusNajmabadi** [reply]
Note: on this point:

> This issue basically breaks the behavior of the return keyword in C#, rendering the behavior of my app unpredicable.

I don't think anyting is broken here.  C# supports the behavior you are seeing.  For example:

```
try
{
    return;
}
finally
{
    // do something
}
```

The code in the finally will execute after the `return` statement.  It is not the nature of C# "that the `return` keyword should immediately stop executing the currently-running method: no passing GO, no collecting $200.".  Indeed, it's vital to C# that thsi not be the case, otherwise things like `finally`s or `using`s would not work.

The docs for postsharp's [LogAttribute](https://doc.postsharp.net/t_postsharp_patterns_diagnostics_logattribute) says:

> Custom attribute that, when applied on a method, causes PostSharp to implement tracing before and after execution of this method. 

So the entire point of this weaver is to ensure that no matter how your method returns it will log the end of it.  If it didn't run code after your return statements, it couldn't accomplish it's core purpose.  So my guess here is that this is simply by design for that attribute and weaver, and the IL it is generating abide by the expectations here of that contract.

**Reply by CyrusNajmabadi** [reply]
>  identify it does not seem endemic to the C-sharp compiler,

Note: even if it was an issue with the C# compiler, the right place to have the discussion/issue would be at dotnet/roslyn.  This repo is for stwardship of the language itself, not the surrounding tools.  Thanks! :)
